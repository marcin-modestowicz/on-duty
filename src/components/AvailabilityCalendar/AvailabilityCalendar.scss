.root {
  .calendar {
    display: flex;
    flex-wrap: wrap;
  }
}
