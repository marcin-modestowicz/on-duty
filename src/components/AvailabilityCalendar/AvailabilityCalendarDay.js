// @flow
import React, { Component } from "react";
import { observer } from "mobx-react";
import classnames from "classnames";
import type Availability, {
  AvailabilityStatus
} from "../../models/Availability";
import {
  AVAILABILITY_STATUSES,
  AVAILABILITY_STATUSES_NAMES
} from "../../models/Availability";
import styles from "./AvailabilityCalendarDay.scss";

type Props = {
  day: number,
  availability: Availability
};

@observer
class AvailabilityCalendarDay extends Component<Props> {
  statuses: AvailabilityStatus[] = [
    AVAILABILITY_STATUSES.FREE,
    AVAILABILITY_STATUSES.KEEN,
    AVAILABILITY_STATUSES.BUSY
  ];

  handleClick = () => {
    this.props.availability.toggle();
  };

  renderDayContent(status: string) {
    return (
      <div key={status} className={classnames(styles.day, styles[status])}>
        <div className={styles.dayName}>Monday</div>
        <div className={styles.dayNum}>{this.props.day}</div>
      </div>
    );
  }

  render() {
    const status = this.props.availability.status;
    const isFreeActive = status === AVAILABILITY_STATUSES.FREE;
    const isKeenActive = status === AVAILABILITY_STATUSES.KEEN;
    const isBusyActive = status === AVAILABILITY_STATUSES.BUSY;

    return (
      <div
        className={classnames(styles.root, {
          [styles.freeActive]: isFreeActive,
          [styles.keenActive]: isKeenActive,
          [styles.busyActive]: isBusyActive
        })}
        role="button"
        title="Click to toggle availability"
        onClick={this.handleClick}
      >
        {this.statuses.map(_status =>
          this.renderDayContent(
            AVAILABILITY_STATUSES_NAMES[_status].toLowerCase()
          )
        )}
      </div>
    );
  }
}

export default AvailabilityCalendarDay;
