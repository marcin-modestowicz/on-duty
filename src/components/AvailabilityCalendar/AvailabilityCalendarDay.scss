@import './colors.scss';

.root {
  position: relative;
  width: 5em;
  height: 5em;
  margin: .25em;

  .day {
    box-sizing: border-box;
    position: absolute;
    border-width: 0 1px 1px 0;
    border-style: solid;
    border-radius: .25em;
    width: 5em;
    height: 5em;
    padding: .75em;
    cursor: pointer;
    user-select: none;
    text-align: center;
    color: white;
    transition: 0.6s;
    transform-style: preserve-3d;
    backface-visibility: hidden;

    .name {
      margin-bottom: .5em;
      font-size: .5em;
    }

    .number {
      font-size: 2em;
      font-weight: bold;

      &:after {
        content: '';
        display: block;
        margin-top: .1em;
        width: 100%;
        height: 1px;
        border-bottom: 1px solid;
      }
    }

    &.free {
      background-color: $free;
      border-color: $free-darker;

      .dayNum:after {
        background-color: $free-darker;
        border-color: $free-lighter;
      }
    }

    &.keen {
      background-color: $keen;
      border-color: $keen-darker;

      .dayNum:after {
        background-color: $keen-darker;
        border-color: $keen-lighter;
      }
    }

    &.busy {
      background-color: $busy;
      border-color: $busy-darker;

      .dayNum:after {
        background-color: $busy-darker;
        border-color: $busy-lighter;
      }
    }
  }

  &.freeActive {
    .free {
      transform: rotateY(0deg);
    }

    .keen {
      transform: rotateY(-180deg);
    }

    .busy {
      transform: rotateY(-180deg);
    }
  }

  &.keenActive {
    .free {
      transform: rotateY(-180deg);
    }

    .keen {
      transform: rotateY(0deg);
    }

    .busy {
      transform: rotateY(-180deg);
    }
  }

  &.busyActive {
    .free {
      transform: rotateY(-180deg);
    }

    .keen {
      transform: rotateY(-180deg);
    }

    .busy {
      transform: rotateY(0deg);
    }
  }
}
