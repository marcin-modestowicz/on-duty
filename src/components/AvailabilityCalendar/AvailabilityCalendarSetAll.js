// @flow
import React, { Component } from "react";
import { observable } from "mobx";
import { observer } from "mobx-react";
import classnames from "classnames";
import {
  type AvailabilityStatus,
  AVAILABILITY_STATUSES,
  AVAILABILITY_STATUSES_NAMES
} from "../../models/Availability";
import styles from "./AvailabilityCalendarSetAll.scss";

type Props = {
  onClick: AvailabilityStatus => void
};

class AvailabilityCalendarSetAll extends Component<Props> {
  handleToggleAll = (status: AvailabilityStatus) => {
    return () => {
      this.props.onClick(status);
    };
  };

  render() {
    return (
      <div className={styles.root}>
        <div className={styles.label}>Set all:</div>
        {Object.keys(AVAILABILITY_STATUSES).map(statusKey => {
          const status = AVAILABILITY_STATUSES[statusKey];
          const statusName = AVAILABILITY_STATUSES_NAMES[status];

          return (
            <div
              key={status}
              role="button"
              className={classnames(
                styles.button,
                styles[statusName.toLowerCase()]
              )}
              title={`Click to set all to ${statusName}`}
              onClick={this.handleToggleAll(status)}
            >
              {statusName}
            </div>
          );
        })}
      </div>
    );
  }
}

export default AvailabilityCalendarSetAll;
