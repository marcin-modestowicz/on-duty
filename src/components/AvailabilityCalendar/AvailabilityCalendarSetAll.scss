@import './colors.scss';

.root {
  display: flex;
  flex-wrap: wrap;

  .label,
  .button {
    box-sizing: border-box;
    width: 5em;
    height: 2.5em;
    padding: .75em;
    margin: .25em;
  }

  .button {
    border-width: 0 1px 1px 0;
    border-style: solid;
    border-radius: .25em;
    cursor: pointer;
    user-select: none;
    text-align: center;
    color: white;

    &:active {
      border-width: 1px 0 0 1px;
    }

    &.free {
      background-color: $free;
      border-color: $free-darker;

      .dayNum:after {
        background-color: $free-darker;
        border-color: $free-lighter;
      }
    }

    &.keen {
      background-color: $keen;
      border-color: $keen-darker;

      .dayNum:after {
        background-color: $keen-darker;
        border-color: $keen-lighter;
      }
    }

    &.busy {
      background-color: $busy;
      border-color: $busy-darker;

      .dayNum:after {
        background-color: $busy-darker;
        border-color: $busy-lighter;
      }
    }
  }
}
