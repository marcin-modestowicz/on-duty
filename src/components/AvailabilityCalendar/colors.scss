$free: #eec452;
$free-lighter: #f7cb52;
$free-darker: #d4ae47;;
$keen: #369984;
$keen-lighter: #43a993;
$keen-darker: #286f60;
$busy: #d32857;
$busy-lighter: #e83a6a;
$busy-darker: #9c1c3f;
