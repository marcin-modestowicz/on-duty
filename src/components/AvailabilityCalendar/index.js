//@flow
export { default } from "./AvailabilityCalendar";
