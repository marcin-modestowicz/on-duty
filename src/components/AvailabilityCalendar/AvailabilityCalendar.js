// @flow
import React, { Component } from "react";
import User from "../../models/User";
import { type AvailabilityCalendar as AvailabilityCalendarModel } from "../../models/Calendar";
import { type AvailabilityStatus } from "../../models/Availability";
import AvailabilityCalendarSetAll from "./AvailabilityCalendarSetAll";
import AvailabilityCalendarDay from "./AvailabilityCalendarDay";
import styles from "./AvailabilityCalendar.scss";

type Props = {
  calendar: AvailabilityCalendarModel
};

class AvailabilityCalendar extends Component<Props> {
  static SetAll = AvailabilityCalendarSetAll;
  static Day = AvailabilityCalendarDay;

  handleToggleAll = (status: AvailabilityStatus) => {
    this.props.calendar.days.forEach(day => {
      day.status = status;
    });
  };

  render() {
    return (
      <div className={styles.root}>
        <AvailabilityCalendar.SetAll onClick={this.handleToggleAll} />
        <div className={styles.calendar}>
          {this.props.calendar.days.map((availability, index) => (
            <AvailabilityCalendar.Day
              key={index}
              day={index + 1}
              availability={availability}
            />
          ))}
        </div>
      </div>
    );
  }
}

export default AvailabilityCalendar;
