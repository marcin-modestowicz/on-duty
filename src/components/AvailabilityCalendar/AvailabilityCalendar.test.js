// @flow
import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import User from "../../models/User";
import { AvailabilityCalendar as AvailabilityCalendarModel } from "../../models/Calendar";
import UserStore from "../../stores/UserStore";
import AvailabilityCalendar from "./AvailabilityCalendar";

describe("AvailabilityCalendar component", () => {
  let RealDate;
  let availabilityCalendar;

  beforeAll(() => {
    RealDate = Date;

    const date = new Date(2017, 9, 1);
    global.Date = function DateMock(...args) {
      return args.length ? new RealDate(...args) : date;
    };

    const calendar = new AvailabilityCalendarModel();
    availabilityCalendar = shallow(
      <AvailabilityCalendar
        calendar={calendar}
        onDayToggle={() => {}}
        onAllDaysStatusSet={() => {}}
      />
    );
  });

  afterAll(() => {
    global.Date = RealDate;
  });

  test("should render", () => {
    expect(toJson(availabilityCalendar)).toMatchSnapshot();
  });
});
