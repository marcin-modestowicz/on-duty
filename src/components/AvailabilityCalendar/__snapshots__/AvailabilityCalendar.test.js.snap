// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvailabilityCalendar component should render 1`] = `
<div
  className="root"
>
  <AvailabilityCalendarSetAll
    onClick={[Function]}
  />
  <div
    className="calendar"
  >
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={1}
      key="0"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={2}
      key="1"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={3}
      key="2"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={4}
      key="3"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={5}
      key="4"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={6}
      key="5"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={7}
      key="6"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={8}
      key="7"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={9}
      key="8"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={10}
      key="9"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={11}
      key="10"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={12}
      key="11"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={13}
      key="12"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={14}
      key="13"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={15}
      key="14"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={16}
      key="15"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={17}
      key="16"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={18}
      key="17"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={19}
      key="18"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={20}
      key="19"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={21}
      key="20"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={22}
      key="21"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={23}
      key="22"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={24}
      key="23"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={25}
      key="24"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={26}
      key="25"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={27}
      key="26"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={28}
      key="27"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={29}
      key="28"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={30}
      key="29"
    />
    <AvailabilityCalendarDay
      availability={
        Availability {
          "status": 0,
        }
      }
      day={31}
      key="30"
    />
  </div>
</div>
`;
